File: src/main/java/com/und/security/RestTokenUtil.java
package com.und.security;

import com.und.common.utils.DateUtils;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.mobile.device.Device;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.io.Serializable;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Component
public class RestTokenUtil implements Serializable {

    private static final long serialVersionUID = -3301605591108950415L;

    static final String CLAIM_KEY_USERNAME = "sub";
    static final String CLAIM_KEY_AUDIENCE = "audience";
    static final String CLAIM_KEY_CREATED = "created";
    static final String CLAIM_KEY_EXPIRED = "exp";

    static final String AUDIENCE_UNKNOWN = "unknown";
    static final String AUDIENCE_WEB = "web";
    static final String AUDIENCE_MOBILE = "mobile";
    static final String AUDIENCE_TABLET = "tablet";

    @Autowired
    private DateUtils dateUtils;

    @Value("${security.expiration}")
    private Long expiration;

    public String getUsernameFromToken(String token, String secret) {
        String username;
        try {
            final Claims claims = getClaimsFromToken(token, secret);
            username = claims.getSubject();
        } catch (Exception e) {
            username = null;
        }
        return username;
    }

    public Date getCreatedDateFromToken(String token, String secret) {
        Date created;
        try {
            final Claims claims = getClaimsFromToken(token, secret);
            created = new Date((Long) claims.get(CLAIM_KEY_CREATED));
        } catch (Exception e) {
            created = null;
        }
        return created;
    }

    public Date getExpirationDateFromToken(String token, String secret) {
        Date expiration;
        try {
            final Claims claims = getClaimsFromToken(token, secret);
            expiration = claims.getExpiration();
        } catch (Exception e) {
            expiration = null;
        }
        return expiration;
    }

    public String getAudienceFromToken(String token, String secret) {
        String audience;
        try {
            final Claims claims = getClaimsFromToken(token, secret);
            audience = (String) claims.get(CLAIM_KEY_AUDIENCE);
        } catch (Exception e) {
            audience = null;
        }
        return audience;
    }

    private Claims getClaimsFromToken(String token, String secret) {
        Claims claims;
        try {
            claims = Jwts.parser()
                    .setSigningKey(secret)
                    .parseClaimsJws(token)
                    .getBody();
        } catch (Exception e) {
            claims = null;
        }
        return claims;
    }

    private Boolean isTokenExpired(String token, String secret) {
        final Date expiration = getExpirationDateFromToken(token, secret);
        return expiration.before(dateUtils.now());
    }

    private Boolean isCreatedBeforeLastPasswordReset(Date created, Date lastPasswordReset) {
        return (lastPasswordReset != null && created.before(lastPasswordReset));
    }

    private String generateAudience(Device device) {
        String audience = AUDIENCE_UNKNOWN;
        if (device.isNormal()) {
            audience = AUDIENCE_WEB;
        } else if (device.isTablet()) {
            audience = AUDIENCE_TABLET;
        } else if (device.isMobile()) {
            audience = AUDIENCE_MOBILE;
        }
        return audience;
    }
    //9654339369 gas agency number
    private Boolean ignoreTokenExpiration(String token, String secret) {
        String audience = getAudienceFromToken(token, secret);
        return (AUDIENCE_TABLET.equals(audience) || AUDIENCE_MOBILE.equals(audience));
    }

    public String generateToken(EventUser userDetails, Device device) {
        Map<String, Object> claims = new HashMap<>();

        claims.put(CLAIM_KEY_USERNAME, userDetails.getUsername());
        claims.put(CLAIM_KEY_AUDIENCE, generateAudience(device));

        final Date createdDate = dateUtils.now();
        claims.put(CLAIM_KEY_CREATED, createdDate);

        return doGenerateToken(claims, userDetails.getSecret());
    }

    private String doGenerateToken(Map<String, Object> claims, String secret) {
        final Date createdDate = (Date) claims.get(CLAIM_KEY_CREATED);
        final Date expirationDate = new Date(createdDate.getTime() + expiration * 1000);

        System.out.println("doGenerateToken " + createdDate);

        return Jwts.builder()
                .setClaims(claims)
                .setExpiration(expirationDate)
                .signWith(SignatureAlgorithm.HS512, secret)
                .compact();
    }

    public Boolean canTokenBeRefreshed(String token, Date lastPasswordReset, String secret) {
        final Date created = getCreatedDateFromToken(token,secret);
        return !isCreatedBeforeLastPasswordReset(created, lastPasswordReset)
                && (!isTokenExpired(token,secret) || ignoreTokenExpiration(token,secret));
    }

    public String refreshToken(String token, String secret) {
        String refreshedToken;
        try {
            final Claims claims = getClaimsFromToken(token,secret);
            claims.put(CLAIM_KEY_CREATED, dateUtils.now());
            refreshedToken = doGenerateToken(claims,secret);
        } catch (Exception e) {
            refreshedToken = null;
        }
        return refreshedToken;
    }

    public Boolean validateToken(String token, UserDetails userDetails) {
        EventUser user = (EventUser) userDetails;
        final String username = getUsernameFromToken(token, user.getSecret());
        final Date created = getCreatedDateFromToken(token,user.getSecret());
        //final Date expiration = getExpirationDateFromToken(token);
        return (
                username.equals(user.getUsername())
                        && !isTokenExpired(token,user.getSecret())
                        && !isCreatedBeforeLastPasswordReset(created, user.getLastPasswordResetDate()));
    }
}

File: src/main/kotlin/com/und/security/RestTokenUtil.kt
package com.und.security

import com.und.common.utils.DateUtils
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.mobile.device.Device
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component
import java.io.Serializable
import java.util.*

@Component
class RestTokenUtil : Serializable {

    @Autowired
    lateinit private var dateUtils: DateUtils

    @Value("\${security.expiration}")
    private var expiration: Long? = 0

    fun getUsernameFromToken(token: String, secret: String): String? {
        val username: String?
        val claims = getClaimsFromToken(token, secret)
        username = claims!!.subject
        return username
    }

    fun getCreatedDateFromToken(token: String, secret: String): Date? {
        val created: Date?
        val claims = getClaimsFromToken(token, secret)
        created = Date(claims!![CLAIM_KEY_CREATED] as Long)

        return created
    }

    fun getExpirationDateFromToken(token: String, secret: String): Date? {
        var expiration: Date?
        try {
            val claims = getClaimsFromToken(token, secret)
            expiration = claims!!.expiration
        } catch (e: Exception) {
            expiration = null
        }

        return expiration
    }

    fun getAudienceFromToken(token: String, secret: String): String? {
        var audience: String?
        try {
            val claims = getClaimsFromToken(token, secret)
            audience = claims!![CLAIM_KEY_AUDIENCE] as String
        } catch (e: Exception) {
            audience = null
        }

        return audience
    }

    private fun getClaimsFromToken(token: String, secret: String): Claims? {
        val claims: Claims? =
                Jwts.parser()
                        .setSigningKey(secret)
                        .parseClaimsJws(token)
                        .body

        return claims
    }

    private fun isTokenExpired(token: String, secret: String): Boolean {
        val expiration = getExpirationDateFromToken(token, secret)
        return expiration!!.before(dateUtils.now())
    }

    private fun isCreatedBeforeLastPasswordReset(created: Date?, lastPasswordReset: Date?): Boolean {
        return lastPasswordReset != null && created!!.before(lastPasswordReset)
    }

    private fun generateAudience(device: Device): String {
       val audience = when {
            device.isNormal -> AUDIENCE_WEB
            device.isMobile -> AUDIENCE_MOBILE
            device.isTablet -> AUDIENCE_TABLET
            else -> AUDIENCE_UNKNOWN
        }
        return audience
    }

    //9654339369 gas agency number
    private fun ignoreTokenExpiration(token: String, secret: String): Boolean {
        val audience = getAudienceFromToken(token, secret)
        return AUDIENCE_TABLET == audience || AUDIENCE_MOBILE == audience
    }

    fun generateToken(userDetails: UndUserDetails, device: Device): String {
        val claims = HashMap<String, Any>()
            claims.put(CLAIM_KEY_USERNAME, userDetails.username)
            claims.put(CLAIM_KEY_AUDIENCE, generateAudience(device))

            val createdDate = dateUtils.now()
            claims.put(CLAIM_KEY_CREATED, createdDate)
            return doGenerateToken(claims, userDetails.secret)
    }

    private fun doGenerateToken(claims: Map<String, Any>, secret: String): String {
        val createdDate = claims[CLAIM_KEY_CREATED] as Date
        val expirationDate = Date(createdDate.time + expiration!! * 1000)

        println("doGenerateToken " + createdDate)

        return Jwts.builder()
                .setClaims(claims)
                .setExpiration(expirationDate)
                .signWith(SignatureAlgorithm.HS512, secret)
                .compact()
    }

    fun canTokenBeRefreshed(token: String, lastPasswordReset: Date, secret: String): Boolean {
        val created = getCreatedDateFromToken(token, secret)
        return !isCreatedBeforeLastPasswordReset(created, lastPasswordReset) && (!isTokenExpired(token, secret) || ignoreTokenExpiration(token, secret))
    }

    fun refreshToken(token: String, secret: String): String {
        val refreshedToken: String
        val claims = getClaimsFromToken(token, secret)
        claims!!.put(CLAIM_KEY_CREATED, dateUtils.now())
        refreshedToken = doGenerateToken(claims, secret)

        return refreshedToken
    }

    fun validateToken(token: String, userDetails: UserDetails): Boolean {
        val user = userDetails as UndUserDetails
        val username = getUsernameFromToken(token, user.secret)
        val created = getCreatedDateFromToken(token, user.secret)
        //final Date expiration = getExpirationDateFromToken(token);
        return username == user.username
                && !isTokenExpired(token, user.secret)
                && !isCreatedBeforeLastPasswordReset(created, user.lastPasswordResetDate)
    }

    companion object {

        private const val serialVersionUID = -3301605591108950415L

        internal val CLAIM_KEY_USERNAME = "sub"
        internal val CLAIM_KEY_AUDIENCE = "audience"
        internal val CLAIM_KEY_CREATED = "created"
        internal val CLAIM_KEY_EXPIRED = "exp"

        internal val AUDIENCE_UNKNOWN = "unknown"
        internal val AUDIENCE_WEB = "web"
        internal val AUDIENCE_MOBILE = "mobile"
        internal val AUDIENCE_TABLET = "tablet"
    }
}